UserInfo/UserInfo/DataLayer/Storage/UsersStorageCoreData.swift

```diff
@@ -0,0 +1,148 @@
+//
+//  UsersStorageCoreData.swift
+//  UserInfo
+//
+//  Created by Oleksandr Savchenko on 21.01.25.
+//
+
+import CoreData
+import OSLog
+
+
+final class UsersStorageCoreData: NSObject, UsersStorage {
+    private static let modelName = "UsersModel"
+    
+    private let container: NSPersistentContainer
+    
+    var managedContext: NSManagedObjectContext {
+        //TODO: add thread safety
+        return self.container.viewContext
+    }
+    
+    override init() {
+        self.container = NSPersistentContainer(name: Self.modelName)
+        self.container.viewContext.mergePolicy = NSMergePolicy(merge: .mergeByPropertyStoreTrumpMergePolicyType)
+        
+        self.container.loadPersistentStores { (storeDescription, error) in
+            Logger.dataStorage.error("Unable to load PersistentStore: \(error)")
+            assert(
+                error == nil,
+                "Unable to load CoreData's Persistent Store. Error: \(error.debugDescription)"
+            )
+        }
+    }
+    
+    func writeUnique(entities: [UserEntity], with token: UUID) async throws(DataStorageError) -> [UserEntity] {
+        do {
+            try await saveToken(token)
+        } catch {
+            Logger.dataStorage.error("Unable to save token: \(error)")
+        }
+        
+        var newUniqueEntities: [UserEntity] = []
+        
+        for entity in entities {
+            if !existing(email: entity.email) {
+                _ = UserEntityMO(context: self.managedContext, with: entity)
+                newUniqueEntities.append(entity)
+            }
+        }
```

Comment:
Это медленный способ работы с массивами, из-за специфики добавления новых значений в динамические массивы. Лучше использовать функциональное программирование, это будет более оптимизированное решение:

```suggestion
        let newUniqueEntities = entities.filter { entity in
            if !existing(email: entity.email) {
                _ = UserEntityMO(context: self.managedContext, with: entity)
                return true
            }
            return false
        }
```